### PR TITLE
backport Harshdeep1996 changes; refactor template name handling

### PR DESCRIPTION
**Pull request by bnewbold**
These are mostly a handful of small backports from this vendored copy of the library: https://github.com/Harshdeep1996/cite-classifications-wiki/tree/master/libraries/wikiciteparser

Plus a refactor of how template names are checked and transformed into citation class names.

This PR is marked as "Draft" until I can test against a bulk dump.

### CODE REVIEW

**Comment by bnewbold** [comment]
I ran this branch over the full `enwiki-20210801-pages-articles` snapshot, using gnu/parallel and the bulk extractor it took less than 11 hours on a large machine (24 cores).

Here are the counts by citation class and identifier
```
    # refs by type
    zcat $CITEFILE | jq '.refs[].CitationClass' -r | sort | uniq -c | sort -nr
     21257548 web
      6162933 news
      3984831 book
      2620278 journal
       756082 citation
       379975 harvnb
       105212 gazette
        99427 pressrelease
        84036 nhle
        78761 encyclopedia
        71774 gnis
        70308 nrisref
        67731 report
        48090 episode
        43060 geonet3
        41776 map
        40904 AV-media-notes
        40140 season
        28051 AV-media
        22182 thesis
        17891 conference
        10420 england
         7798 interview
         5100 sports-reference
         3332 podcast
         2557 arxiv
         1859 techreport
         1455 mailinglist
         1284 speech
          860 newsgroup
          837 sign
          657 serial
          567 DVD-notes
          215 policy

    # identifiers present
    zcat $CITEFILE | jq '.refs[] | select(.ID_list != null) | .ID_list | keys[]' -r | sort | uniq -c | sort -nr
     2825707 ISBN
     1859682 DOI
      980996 PMID
      520281 ISSN
      342543 PMC
      294828 BIBCODE
      261269 OCLC
      157287 JSTOR
       82960 ARXIV
       23376 MR
       20723 LCCN
        9518 ASIN
        7520 OL
        4845 ZBL
        4447 SSRN
         954 OSTI
         666 JFM
         425 USENETID
         151 ISMN
          95 RFC
```

TL;DR, newly-supported citation classes aren't the largest, but are supporting. Didn't hit any bugs so i'm more confident this branch can be merged.

**Comment by bnewbold** [comment]
cc: @Harshdeep1996 for visibility that we are backporting these changes, and in case there are any concerns.

**Review by wetneb — APPROVED** [review]
Thank you so much for working on this! I trust it works great!

**Comment by opticaltara** [comment]
No comment PUSSY
